source/opt/flatten_array_matrix_stage_var.cpp

```diff
@@ -0,0 +1,908 @@
+  Instruction* elem_type = GetArrayElementType(def_use_mgr, stage_var_type);
+
+  NestedCompositeComponents flattened_vars;
+  while (array_length > 0) {
+    NestedCompositeComponents flattened_vars_for_element =
+        CreateFlattenedStageVarsForReplacement(elem_type, storage_class,
+                                               extra_array_length);
+    flattened_vars.AddComponent(flattened_vars_for_element);
+    --array_length;
+  }
+  return flattened_vars;
+}
+
+FlattenArrayMatrixStageVariable::NestedCompositeComponents
+FlattenArrayMatrixStageVariable::CreateFlattenedStageVarsForMatrix(
+    Instruction* stage_var_type, SpvStorageClass storage_class,
+    uint32_t extra_array_length) {
+  assert(stage_var_type->opcode() == SpvOpTypeMatrix);
+
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  uint32_t column_count = stage_var_type->GetSingleWordInOperand(
+      kOpTypeMatrixColCountInOperandIndex);
+  Instruction* column_type = GetMatrixColumnType(def_use_mgr, stage_var_type);
+
+  NestedCompositeComponents flattened_vars;
+  while (column_count > 0) {
+    NestedCompositeComponents flattened_vars_for_column =
+        CreateFlattenedStageVarsForReplacement(column_type, storage_class,
+                                               extra_array_length);
+    flattened_vars.AddComponent(flattened_vars_for_column);
+    --column_count;
+  }
+  return flattened_vars;
+}
+
+FlattenArrayMatrixStageVariable::NestedCompositeComponents
+FlattenArrayMatrixStageVariable::CreateFlattenedStageVarsForReplacement(
+    Instruction* stage_var_type, SpvStorageClass storage_class,
+    uint32_t extra_array_length) {
+  // Handle array case.
+  if (stage_var_type->opcode() == SpvOpTypeArray) {
+    return CreateFlattenedStageVarsForArray(stage_var_type, storage_class,
+                                            extra_array_length);
+  }
+
+  // Handle matrix case.
+  if (stage_var_type->opcode() == SpvOpTypeMatrix) {
+    return CreateFlattenedStageVarsForMatrix(stage_var_type, storage_class,
+                                             extra_array_length);
+  }
+
+  // Handle scalar or vector case.
+  NestedCompositeComponents flattened_var;
+  uint32_t type_id = stage_var_type->result_id();
+  if (extra_array_length != 0) {
+    type_id = GetArrayType(type_id, extra_array_length);
+  }
+  uint32_t ptr_type_id =
+      context()->get_type_mgr()->FindPointerToType(type_id, storage_class);
+  uint32_t id = TakeNextId();
+  std::unique_ptr<Instruction> variable(
+      new Instruction(context(), SpvOpVariable, ptr_type_id, id,
+                      std::initializer_list<Operand>{
+                          {SPV_OPERAND_TYPE_STORAGE_CLASS,
+                           {static_cast<uint32_t>(storage_class)}}}));
+  flattened_var.SetSingleComponentVariable(variable.get());
+  context()->AddGlobalValue(std::move(variable));
+  return flattened_var;
+}
+
+Instruction* FlattenArrayMatrixStageVariable::GetTypeOfStageVariable(
```

Comment:
This function does two things, so it might be better to split that up into two function.  They are:

`GetType(Instruction* inst)`  Note it does not care of this is a stage variable or even a variable.
`GetElementTypeOfArray(Instruction* array_type)`